### PR TITLE
Implement missing API for vscode.typescript-language-features@1.62.3

### LINKED ISSUE #10819: Built-in TypeScript doesn't understand Theia
### Feature Description:

The version of TypeScript built into the example application doesn't include support for the `override` keyword or the `noImplicitOverride` flag. We should make sure that the example application can be used smoothly for writing Theia itself.

- [ ] #11036
- [x] TextDocumentChangeReason

### LINKED ISSUE #10979: vscode: support `createLanguageStatusItem` API
#### Feature Description:

The goal is to support the `createLanguageStatusItem` VS Code API:

image

For reference [`vscode-builtin-typescript-language-features@1.62.3`](https://open-vsx.org/extension/vscode/typescript-language-features) uses this API and it fails to activate.

### LINKED ISSUE #11036: Support `vscode.LanguageStatusItem`
### Feature Description:

Implement the types and status-bar logic necessary to support VSCode's `LanguageStatusItem`.

Required for recent TS versions.

### LINKED ISSUE #11129: [vscode] Support optional property `MarkdownString#supportThemeIcons`
### Feature Description:
Support optionally specifying whether a `MarkdownString` supports `ThemeIcons` via the constructor or property `supportThemeIcons`, see https://code.visualstudio.com/api/references/vscode-api#MarkdownString

See also #11128

### PR DESCRIPTION
**Pull request by colin-grant-work**
#### What it does

Fixes #11129 by updating the `MarkdownString` implementation available to plugins.
Fixes #10819, fixes #10979, fixes #11036

- [x] [Registered CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=24086)

This PR implements missing API required by recent versions of the TypeScript plugin. In particular:

 - `LanguageStatusItem` and related API (#10979 & #11036)
 > This required a significant reworking of the `StatusBar`, and much of this is copied or adapted from VSCode.
 > - Support for affinity between two items
 > - Support for MarkdownStrings and HTML elements in hovers
 > - Improvement in MarkdownString types and implementations
 - `TextDocumentChangeReason`

##### TS Plugin Activation, Language Server
1. Download the [TS plugin](https://open-vsx.org/extension/vscode/typescript) and [TS language features plugin](https://open-vsx.org/extension/vscode/typescript-language-features) and replace the corresponding entries in your `theia/plugins` folder.
2. Start the application with Theia as workspace and open a TypeScript file.
3. Observe that you do not see any toasts about the plugin failing to activate.
4. Use some new features, e.g. find an `override` modifier and `ctrl+click` to go to the super method. Observe that `override` is not treated as a syntax error.

##### Language Status Item (and HTML-element tooltips in StatusBar)
1. With a TS file open, observe that you get a language status item (`{}`) next to the language indicator. 
2. Hover over that, and observe that two items are available.
3. Clicking the links should execute the associated commands to select TS version or open the tsconfig file for the current editor.
4. Clicking the pins should add / remove that command to / from the status bar as an independent item, and clicking that item should execute the command.
5. The pinned items should appear and disappear according to the language of the current editor.

##### Markdown Rendering (MarkdownString in tooltips in StatusBar)
0. Put [markdown-renderer-test-0.0.1.zip](https://github.com/eclipse-theia/theia/files/8572871/markdown-renderer-test-0.0.1.zip) in your plugins folder and start the application.
1. Start the application.
2. Run the command `VSCode API: Show status item with Markdown tooltip`
3. A status item should appear on the left side with the text 'Check my tooltip'
4. Hover over that item, and you should see 
image

The gear should spin, and if you have an editor open as the current widget, clicking the link should trigger the 'save-as' action for that editor.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by planger** [comment]
@colin-grant-work I saw that there's a playwright test that consistently fails with this PR.
```
[playwright]     > 43 |         expect(await notificationIndicator.isVisible()).toBe(true);
[playwright]          |                                                         ^
...
[playwright]   1 failed
[playwright]     ../../src/tests/theia-status-bar.test.ts:38:9 › Theia Status Bar › should contain status bar elements 
[playwright]   1 skipped
[playwright]   49 passed (2m)
```
Looks like the selector for the notification indicator doesn't work anymore.
Just wanted to offer to look into it, if you want. Feel free to skip this test and merge this PR, once it is ready, and create an issue for us to fix and re-enable the test.

**Comment by msujew** [comment]
I'll look into this once I'm back from vacation in mid May.

**Comment by colin-grant-work** [comment]
@jfaltermeier, thanks for checking. I implemented a service that renders tooltips on status bar items as HTML elements rather than using the `title` property, but it looks like some of the machinery to determine whether a given status bar item is visible checks the `title`. At the moment, we don't put much other metadata on those elements, but we could for example add an `id` property that would make it easy to pick out one status bar item among the rest?

**Comment by planger** [comment]
@colin-grant-work Thanks for the feedback!

I think we currently use either icons or titles as selectors for notifications or status bar items in general. Originally, because there wasn't other metadata and it's the closest to what the user would use to identify a status bar item.

But `id` or `data-`attributes would be fine too and it would be very low effort to change that in the page objects.
If you could add those additional attributes, that'd be great. Otherwise I'm happy to add them too, when I fix up the page objects. As you prefer.

Thanks!

**Comment by vince-fugnitto** [comment]
The [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=24086) opened for this pull-request has been approved :+1:

File: packages/editor/src/common/paths.ts [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+// *****************************************************************************
+
+// copied from https://github.com/Microsoft/vscode/blob/bf7ac9201e7a7d01741d4e6e64b5dc9f3197d97b/src/vs/base/common/paths.ts
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+/* eslint-disable no-void */
+/* eslint-disable no-null/no-null */
+'use strict';
```
[the remark this answers is not in the document]
Comment (colin-grant-work) [reply packages/editor/src/common/paths.ts:24]: I agree, but I also didn't add it - this was just copied from `callhierarchy`. Perhaps it should be moved into `core`? There are quite a number of things that were copied from VSCode into `plugin-ext` (or somewhere else for use in `plugin-ext`) that really basically belong in `core`, but I'm not sure whether we now want to copy them into core. Some examples I've run across recently are things like `UntitledResource` and much of the code to handle `MarkdownString`s. What do you think @vince-fugnitto, re: whether to move this clear into `core` or leave it here?

**Comment by colin-grant-work** [comment]
@planger, I have added an `id` to status bar elements and [reworked](https://github.com/eclipse-theia/theia/pull/11083/commits/3d72edf15ea000dbc62be0687351ac8d8e648b5d) the corresponding page objects. I ended up modifying them rather a lot because finding things by id is much easier than by icon or by title, but it may be desirable to retain some of the old icon-related functionality. Let me know what you think.

File: packages/core/src/browser/status-bar/status-bar-types.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { MarkdownString } from '../../common/markdown-rendering/markdown-string';
+import { AccessibilityInformation } from '../../common/accessibility';
+
+export interface StatusBarEntry {
+    /**
+     * For icons we use octicons and fontawesome icons. octicons take precedence over fontawesome. Get more information and the class names
```
[the remark this answers is not in the document]
Comment (colin-grant-work) [reply packages/core/src/browser/status-bar/status-bar-types.ts]: I've updated this to reflect the priority of codicons, and also added handling for cases of `fa-` prefixed icons.

File: packages/core/src/browser/status-bar/status-bar-hover-manager.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+// *****************************************************************************
+
+import { inject, injectable } from 'inversify';
+import { Disposable, DisposableCollection, disposableTimeout, isOSX } from '../../common';
+import { MarkdownString } from '../../common/markdown-rendering/markdown-string';
+import { MarkdownRenderer, MarkdownRendererFactory } from '../markdown-rendering/markdown-renderer';
+import { PreferenceService } from '../preferences';
+
+@injectable()
+export class StatusBarHoverManager {
```
[the remark this answers is not in the document]
Comment (colin-grant-work) [reply packages/core/src/browser/status-bar/status-bar-hover-manager.ts:24]: I agree with this suggestion, but I think I'd like to defer it to a separate PR. I think creating a fully general hover service supporting a wide range of relevant options would take some work and some additional testing, and this PR is already quite large. What do you think?

**Comment by planger** [comment]
> @planger, I have added an `id` to status bar elements and [reworked](https://github.com/eclipse-theia/theia/pull/11083/commits/3d72edf15ea000dbc62be0687351ac8d8e648b5d) the corresponding page objects. I ended up modifying them rather a lot because finding things by id is much easier than by icon or by title, but it may be desirable to retain some of the old icon-related functionality. Let me know what you think.

Very nice! Thank you!

You are certainly right, working with IDs makes the code much leaner. I think the use case for selecting by icons is rather uncommon for such rather static core elements, like status bar items. So I think only using the ID is completely fine for now.

Thanks again!

File: packages/core/src/browser/status-bar/status-bar-hover-manager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (msujew) [reply packages/core/src/browser/status-bar/status-bar-hover-manager.ts:24]: Alright, yeah I agree 👍

File: packages/editor/src/browser/language-status/editor-language-status-service.ts [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import '../../../src/browser/language-status/editor-language-status.css';
```
Comment (vince-fugnitto) [inline packages/editor/src/browser/language-status/editor-language-status-service.ts:17]: minor: I believe we generally import css through `*-modules.ts`.

File: packages/editor/src/browser/language-status/editor-language-status.css [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+.hover-language-status>.element {
+    display: flex;
+    justify-content: space-between;
+    vertical-align: middle;
+    flex-grow: 100;
+}
+
+.hover-language-status>.element>.left>.detail:not(:empty)::before {
+    /* endash */
```
Comment (vince-fugnitto) [inline packages/editor/src/browser/language-status/editor-language-status.css:55]: I might be wrong but isn't it spelled like so: ;)

```suggestion
    /* en dash */
```

File: packages/editor/src/common/language-selector.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+import { match as matchGlobPattern } from '@theia/core/lib/common/glob';
```
Comment (vince-fugnitto) [inline packages/editor/src/common/language-selector.ts:16]: minor: newline.

**Review by msujew — APPROVED** [review]
Thanks, the changes look good to me!

- [x] The updated versions of the typescript extensions start as expected
- [x] Markdown rendered hovers behave as in vscode
- [x] Status bar elements continue to work as expected

File: packages/editor/src/browser/language-status/editor-language-status.css (hunk 5 above)
Comment (colin-grant-work) [reply packages/editor/src/browser/language-status/editor-language-status.css:55]: Compromise in `en-dash`?

**Review by vince-fugnitto — APPROVED** [review]
The changes work well for me 👍 

I confirmed that:
- [x] existing statusbar items contributed by the framework look and work as expected
- [x] the typescript-language-server builtin works without errors
    - [x] the `{ }` statusbar item looks good and works
    - [x] the `{ }` pin/unpin functionality works
    - [x] the `{ }` config file works
- [x] the contributed statusbar item works in markdown, icons are properly displayed

**Pull request merged**